### PR TITLE
Fix #2

### LINKED ISSUE #2: Failed authentication
When authentication fails
```
begin
  client.authenticate(user, pass)
rescue ex
  puts ex.message
end
```
got unhandled exception
```
bind failed with InvalidCredentials: Simple Bind Failed: NT_STATUS_LOGON_FAILURE
Unhandled exception in spawn: Channel is closed (Channel::ClosedError)
  from /usr/lib/crystal/channel.cr:228:9 in 'send'
  from /usr/lib/crystal/log/dispatch.cr:55:7 in 'dispatch'
  from /usr/lib/crystal/log/backend.cr:24:5 in 'dispatch'
  from /usr/lib/crystal/log/log.cr:36:3 in 'process!'
  from lib/ldap/src/ldap/client.cr:27:13 in '->'
  from /usr/lib/crystal/primitives.cr:255:3 in 'run'
  from /usr/lib/crystal/fiber.cr:92:34 in '->'
  from ???
```

### PR DESCRIPTION
**Pull request by mbab**
Hope is valid.

### CODE REVIEW

**Comment by stakach** [comment]
I am fairly certain that is is unrelated to the issue #2 
The issue seems to be with logging configuration and the Log dispatcher configuration

**Comment by n-rodriguez** [comment]
Dropping by while auditing this repo toward a `1.0` ([#6](https://github.com/spider-gazelle/crystal-ldap/issues/6)) — I dug into this one because it touches the error model. Sharing the analysis in case it helps close the thread.

**TL;DR:** @stakach's instinct in [the comment above](https://github.com/spider-gazelle/crystal-ldap/pull/3#issuecomment-801465952) is right — the crash in #2 is a logging/dispatch-at-shutdown problem, not the `close` call. Removing `close` makes the symptom *disappear in this one path* but doesn't fix the cause, and adds a leak.

### What actually happens in #2

The unhandled `Channel::ClosedError` is raised inside the spawned `process!` fiber, not in `authenticate`:

```crystal
# client.cr — process!
while !socket.closed?
  data = socket.read_bytes(ASN1::BER)   # close() races here on failed bind
  parse_response data
end
rescue IO::Error
  @mutex.synchronize { close }
rescue e
  Log.error(exception: e) { e.message }  # <-- fires during shutdown
```

1. `authenticate` raises `AuthError` → its `rescue` calls `close`.
2. `close` shuts the socket while `process!` is blocked in `read_bytes`. On a socket closed mid-read, `bindata` can raise something **other** than `IO::Error` (BER decode on truncated/EOF data) → it lands in the generic `rescue e` → `Log.error`.
3. By then the program is tearing down and `Log`'s dispatcher channel is already closed → `Log.error` → `Channel::ClosedError`, unhandled in the fiber.

So the trigger is `close` racing the read **into the `Log.error` path** — removing `close` just sidesteps that one trigger.

### Why this patch isn't the fix

- **Doesn't address the cause:** any other non-`IO::Error` read error still hits `Log.error` and can crash the same way at shutdown. Failed-auth is just one way to reach it.
- **Introduces a leak:** after a failed bind the socket is no longer closed, leaving a half-open connection and a live `process!` fiber.

### Suggested direction

Treat "socket closed locally" as a clean shutdown rather than an error to log, and don't let the read fiber raise unhandled:

```crystal
rescue IO::Error
  @mutex.synchronize { close }
rescue e
  Log.error(exception: e) { e.message } unless @socket.closed?  # don't log a deliberate close
  @mutex.synchronize do
    @requests.values.each(&.reject(e))
    @requests.clear
    close
  end
```

Whether `authenticate` *should* `close` on a failed bind is a separate, legitimate question (per RFC 4511 a failed bind leaves the connection in an anonymous, reusable state) — but that's worth deciding explicitly, not by dropping the line to mask the crash.

Happy to fold a regression spec that reproduces the failed-auth shutdown crash into the Tier-0 / error-model work in #6 if that's useful. 🙏

**Comment by n-rodriguez** [comment]
Followed up with the root-cause fix in **#9** (closes #2), building on @stakach's logging/dispatch diagnosis here.

Rather than dropping the `close` from `#authenticate` (which masks the trigger but leaks the connection + read fiber), #9 stops `process!` from logging a *deliberate* teardown as an error: `Log.error(...) unless @socket.closed?`. So `#authenticate` keeps closing, and the `Channel::ClosedError`-at-shutdown path is gone. Comes with a spec that reproduces the failed-bind teardown and asserts no `:error` is logged.

Thanks @mbab for the original report and repro — that's what made it pin-pointable.
